### ISSUE TITLE
[Feature request] Save file locally on web server

### ISSUE THREAD

**Issue by Rurik19** [antelle/keeweb#15]
Can you add for web version ability to save file right on server ?

**Comment by antelle** [comment]
Web version is a pure client app, it will not require a server for it. To add an ability save and load to own server, I'll add an option to sync with URL via HTTP PUT and GET, this will be added as an option when storage providers ux and functional (onedrive, gdrive, ...) are implemented.

**Comment by imraro** [comment]
>...when storage providers ux and functional (onedrive, gdrive, ...

...[OwnCloud](https://github.com/owncloud), [Syncthing](https://github.com/syncthing/syncthing),...  ))

**Comment by 3ruce** [comment]
+1 this woudl be excellent

**Comment by antelle** [comment]
Which protocol would you prefer? WebDAV? How auth should be implemented for better security?

**Comment by 3ruce** [comment]
WebDAV would be great. I'm not sure how auth should be implemented though but the best server environment for me would be cpanel/shared hosting if that helps? 

Given the excellent 'single file' approach, being able to get the app running in a shared hosting environment via webdav or similar would be excellent!

**Comment by Rurik19** [comment]
About authentication: I think basic http auth (for self-hosted webdav, like me..) and OAuth for many web services, like Yandex Disk.

**Comment by 3ruce** [comment]
Basic http for me...

**Comment by danielfie** [comment]
+1 I'm looking forward to it. WebDav is ok :)

**Comment by yusf** [comment]
[remoteStorage](https://remotestorage.io/) would be another wished for storage protocol.

**Comment by fthiery** [comment]
+1 WebDav would be awesome (but the desktop app already is; thanks for building a clipperz killer !) !

**Comment by antelle** [comment]
So, v1.1 will support WebDav.
Most likely, webdav server will have to match some more strict requirements (like enabled ETag and CORS support).
I'm still investigating authentication possibilities and value.

**Comment by 3ruce** [comment]
1.1 for WebDav sounds great. FYI, I asked my shared hosting provider about ETag and CORS and this is what they said...

>>>
CORS is already supported, because at the server level, what it requires is mod_headers, which is a standard part of a server build's compile. ETag, however, is not: for PCI compliance reasons, that is disabled on the servers, along with other apache-related items like server signatures and trace enabling.
<<<

Any chance of just using CORS?

**Comment by antelle** [comment]
I'm thinking about transational update: we need some mechanism to upload file on server only if it matches the version which war retrieved. This could be done, for example, with ETag. Do you know any replacement which is more widely used?
Without ability of transactional upload, there will be risk of rewriting files, and conflicts will not be resolved correctly. Theoretically, we _can_ use just CORS, but there will be a risk of misbehaving sync.

**Comment by 3ruce** [comment]
hmmm I was just thinking about using a single webdav file/database, I hadn't thought too much about syncing... check out the comments on 'last modified time' below, maybe that's an easier approach?

http://webmasters.stackexchange.com/questions/3650/when-to-use-and-not-to-use-etags

**Comment by antelle** [comment]
From the question on stackexchange which you have just posted:
> Put simply: you hit edit on a page in your CMS, your friend hits edit on a page in your CMS, your friend performs their edit and hits save and finally you hit save - without an ETag or Content-MD5 HTTP header you would need to reinvent the wheel to prevent problems occuring

That's exactly our case. So, without ETags everything will work but once you open the file from two computer and modify if offline, there will be no guarantees. 
There's one more option which I see: if ETags are not enabled, make HEAD request (or GET with If-Modified-Since header set) before update and get last modified time: this is not atomic and vulnerable to race condition but better than just uploading.

**Comment by fthiery** [comment]
So what would happen in case of conflict ? Would the software download the new, merge changes and upload ?

**Comment by antelle** [comment]
The upload will look like:
1. if ETags are enabled - just upload with ETag set; if it's not matched, download, merge and try to upload again
2. if they are not enabled - before upload make additional request to check modifications, and if the file is modified, download, merge, repeat again
While option 1 is consistent, 2 is not: the file could be modified between requests (race condition).
The algorithm described above (option 1) is already implemented for Dropbox sync (they provide file rev).

**Comment by 3ruce** [comment]
Having both option 1 and 2 avaliable makes for more use cases as far as I can tell?

**Comment by antelle** [comment]
I think so as well. Additionally, we can show a warning when ETags are now available and we're using option 2 (it's hard to explain to the user though)

**Comment by 3ruce** [comment]
Yeah, maybe some sort of config option/show only once type message might help users? My use case is for a family member. I want to stick this on a website somewhere and send them a master pass and then they no longer need to bother me with reminders for their own passwords, hence needing something simple(ish)!

**Comment by michaelkuty** [comment]
:+1: for webdav (owncloud) and server side serving

**Comment by rtauxe** [comment]
WebDav with OwnCloud is a very good idea !
Do you know approximately when will the v1.1 available ?

**Comment by antelle** [comment]
Q2 2016, most probably in May
For now, I'm working on general storage providers support

**Comment by rtauxe** [comment]
Thank you for you answer. I'm waiting for it :)

**Comment by antelle** [comment]
First WebDAV implementation is done, closing this as a duplicate of #14, to keep it as single discussion thread.
